contracts/BSCValidatorSet.sol

```diff
@@ -344,6 +352,46 @@ contract BSCValidatorSet is IBSCValidatorSet, System, IParamSubscriber, IApplica
     return CODE_OK;
   }
 
+  function shuffle(address[] memory validators, uint256 epochNumber, uint startIdx, uint offset, uint limit, uint modNumber) internal pure {
+    for (uint i = 0; i<limit; i++) {
+      uint random = uint(keccak256(abi.encodePacked(epochNumber, startIdx+i))) % modNumber;
+      if ( (startIdx+i) != (offset+random) ) {
+        address tmp = validators[startIdx+i];
+        validators[startIdx+i] = validators[offset+random];
+        validators[offset+random] = tmp;
+      }
+    }
+  }
+
+  function getMiningValidators(uint256 epochNumber) public view returns(address[] memory) {
+    uint256 _maxNumOfWorkingCandidates = maxNumOfWorkingCandidates;
+    uint256 _numOfCabinets = numOfCabinets;
+    if (_maxNumOfWorkingCandidates == 0 ){
+      _maxNumOfWorkingCandidates = INIT_MAX_NUM_OF_WORKING_CANDIDATES;
+    }
+    if (_numOfCabinets == 0 ){
+      _numOfCabinets = INIT_NUM_OF_CABINETS;
+    }
+
+    address[] memory validators = getValidators();
```

Comment:
Please make sure the return of `getValidators` is sorted by voting power.

Reply:
I have double checked it. It is not strickly in order...     Please search `It is ok that the validatorSet is not in order.` in the repo.

Reply:
I changed the logic of the felony function to ensure the order of the currentValidatorSet , please check